amazon/ion/ioncmodule.c

```diff
@@ -878,125 +877,14 @@ static PyObject* ionc_get_timestamp_precision(int precision) {
-                PyDict_SetItemString(timestamp_args, "fractional_seconds", py_fractional_seconds);
-                Py_DECREF(py_fractional_seconds);
-                Py_DECREF(py_dec_str);
-            } else {
-                decQuadScaleB(&fraction, &fraction, decQuadFromInt32(&tmp, MICROSECOND_DIGITS), &dec_context);
-                int32_t microsecond = decQuadToInt32Exact(&fraction, &dec_context, DEC_ROUND_DOWN);
-
-                if (decContextTestStatus(&dec_context, DEC_Inexact)) {
-                    // This means the fractional component is not [0, 1) or has more than microsecond precision.
-                    decContextClearStatus(&dec_context, DEC_Inexact);
-                }
-
-                PyObject* py_microsecond = PyLong_FromLong(microsecond);
-                PyObject* py_fractional_precision = PyLong_FromLong(fractional_precision);
-                PyDict_SetItemString(timestamp_args, "microsecond", py_microsecond);
-                PyDict_SetItemString(timestamp_args, "fractional_precision", py_fractional_precision);
-                Py_DECREF(py_microsecond);
-                Py_DECREF(py_fractional_precision);
-            }
-        }
-        case ION_TS_SEC:
-        {
-            PyObject* temp_seconds = PyLong_FromLong(timestamp_value.seconds);
-            PyDict_SetItemString(timestamp_args, "second", temp_seconds);
-            Py_DECREF(temp_seconds);
-        }
-        case ION_TS_MIN:
-        {
-            PyObject* temp_minutes = PyLong_FromLong(timestamp_value.minutes);
-            PyObject* temp_hours = PyLong_FromLong(timestamp_value.hours);
-
-            PyDict_SetItemString(timestamp_args, "minute", temp_minutes);
-            PyDict_SetItemString(timestamp_args, "hour",  temp_hours);
+iERR ionc_read_timestamp(hREADER hreader, PyObject **timestamp_out) {
+   iENTER;
+   ION_TIMESTAMP timestamp_value = {0};
 
-            Py_DECREF(temp_minutes);
-            Py_DECREF(temp_hours);
-        }
-        case ION_TS_DAY:
-        {
-            PyObject* temp_day = PyLong_FromLong(timestamp_value.day);
-            PyDict_SetItemString(timestamp_args, "day", temp_day);
-            Py_DECREF(temp_day);
-        }
-        case ION_TS_MONTH:
-        {   PyObject* temp_month = PyLong_FromLong(timestamp_value.month);
-            PyDict_SetItemString(timestamp_args, "month", temp_month);
-            Py_DECREF(temp_month);
-        }
-        case ION_TS_YEAR:
-        {
-            PyObject* temp_year = PyLong_FromLong(timestamp_value.year);
-            PyDict_SetItemString(timestamp_args, "year", temp_year);
-            Py_DECREF(temp_year);
-            break;
-        }
-    }
-    *timestamp_out = PyObject_Call(_py_timestamp_constructor, PyTuple_New(0), timestamp_args);
+   IONCHECK(ion_reader_read_timestamp(hreader, &timestamp_value));
 
+   *timestamp_out = IonTimestamp_FromTimestamp(&timestamp_value);
```

Comment:
That's a good consistency callout. I had originally intended to name the python type IonTimestamp, which lead to the function name since IonTimestamp_FromIonTimestamp would be misleading. But, now that I'm looking at it again, the type should be either IonPyTimestamp, or Timestamp (from amazon.ion.core). I'll probably rename it to IonPyTimestamp, to take the place of the simple_types implementation. Then the functions would become `IonPyTimestamp_*`

Reply:
I'm not sure I follow exactly, do you mean creating a wrapper for this implementation?

If I were to implement a native type, that implemented the PyIon* pattern for this Timestamp implementation, I'd expect the performance of it compared to the bare Timestamp class to be near identical. The reason I'd expect that, is that the PyTuple for annotations is created whether we emit bare values or not. In the native implementation, storing that tuple would literally be a pointer assignment. Nothing that the wrapper types provide, would take significant time in the native implementation.

Implementing the wrapper in python, like the current types use, however, has issues. As far as I can tell, with the python value implementations right now, each time we create a wrapped value, we're essentially creating the bare value twice. The first is instantiated when we determine what value type we're looking at in the ion reader. Then, once we have that bare value, if we want to return a wrapped value we call the wrapped value's constructor, providing the existing bare value. Since the wrapped value subclasses the bare value, the wrapped value's constructor needs to call the bare value's constructor to allocate and initialize the object. This re-runs the bare value's constructor, using the already created bare value as an argument.

I confirmed the behavior, and the performance impact by profiling a modified version of my test harness that emitted wrapped values only for symbols. The results showed that creating the initial bare value for all symbols in our log dataset, contributed 2.8% to the total profile samples; creating the wrapped values contributed an additional 6.3%. This may seem like a small percentage for the bare symbols, since the log dataset does have a significant number of symbols, but it's worth noting that this is just creating a python string or long, and instantiating the `SymbolToken`. The instantiation contributes 91% of the samples for `ion_string_to_py_symboltoken`, whereas creating the python string which includes validating the UTF-8 data, contributes 5.9%. The `SymbolToken` construction should be, loosely, equivalent to an allocation, and tuple creation, however the tuple's new only accounts for 11% of `ion_string_to_py_symboltoken`'s samples, and the allocation is included in that. So it would seem, that 80% of the samples contributing to creating the bare value is overhead associated with creating the bare value through the interpreter.

To quantify the "through the interpreter" a little more, 52% of `ion_string_to_py_symboltoken`, is contributed by `_PyEval_EvalFrameDefault`, which handles python instruction processing. 17% of samples are contributed by that function itself (not function calls from it).

IMO, the bare value functionality definitely helps with the current python value implementations, but I don't see any major gains for following the pattern for native types (unless I'm missing something), and certainly wouldn't want to implement a python type to wrap the native type.

We could probably refactor the `ionc_read_value` function to not create the bare value unless we know we want a bare value, and create only a wrapped value otherwise. That seems like it could cut ~1/3-1/2 of the cost of using wrapped values.

Reply:
> If I were to implement a native type, that implemented the PyIon* pattern for this Timestamp implementation, I'd expect the performance of it compared to the bare Timestamp class to be near identical.

Agreed.

> This re-runs the bare value's constructor, using the already created bare value as an argument.

Yep.

> We could probably refactor the ionc_read_value function to not create the bare value unless we know we want a bare value, and create only a wrapped value otherwise. That seems like it could cut ~1/3-1/2 of the cost of using wrapped values.

For some... for built-ins, like String and Int you need to bridge from c to Python objects at some point anyway. For collections, I've been refactoring those to avoid copying when possible.

> I'm not sure I follow exactly, do you mean creating a wrapper for this implementation?

I was just responding to your ambivalence "the type should be either IonPyTimestamp, or Timestamp". I would like to see _just_ a `Timestamp`, that may be wrapped by an `IonPyTimestamp`. I agree with your concern about copying but I believe if folks want performance they should just use 'bare values' and then it's only the timestamps that have annotations that need to be wrapped. We can also mitigate the perf hit for that case by having no logic in the constructors.